In [ ]:
import random
import statistics as stat
from simple_markov_chain_lib import markov_chain

random.seed(2018)
p = 1/6
init_probs = {1: 1.0}
markov_table = {
    1: {2: 1.},
    2: {2: 2/3, 3: 1/3},
    3: {1: p, 2: 1-p}
}

mc = markov_chain(markov_table, init_probs)

M = 50       # Πλήθος εκτιμήσεων
N = 200      # Πλήθος δειγμάτων ανά εκτίμηση
steps = 20

estimates = []

for k in range(M):
    counter = 0
    for i in range(N):
        mc.start()
        for j in range(steps):
            mc.move()
        if mc.running_state == 1:
            counter += 1

    phat = counter / N
    estimates.append(phat)

mean_est = stat.mean(estimates)
var_est = stat.variance(estimates)

print(f"--- Αποτελέσματα για N = {N} ---")
print(f"Δειγματική Μέση Τιμή (Mean): {mean_est:.5f}")
print(f"Δειγματική Διασπορά (Variance): {var_est:.5f}")

--- Αποτελέσματα για N = 200 ---
Δειγματική Μέση Τιμή (Mean): 0.03890
Δειγματική Διασπορά (Variance): 0.00012


In [ ]:
random.seed(2018)
p = 1/6
init_probs = {1: 1.0}
markov_table = {
    1: {2: 1.},
    2: {2: 2/3, 3: 1/3},
    3: {1: p, 2: 1-p}
}

mc = markov_chain(markov_table, init_probs)

M = 50       # Πλήθος εκτιμήσεων
N = 20000     # Πλήθος δειγμάτων ανά εκτίμηση
steps = 20

estimates = []

for k in range(M):
    counter = 0
    for i in range(N):
        mc.start()
        for j in range(steps):
            mc.move()
        if mc.running_state == 1:
            counter += 1

    phat = counter / N
    estimates.append(phat)

mean_est = stat.mean(estimates)
var_est = stat.variance(estimates)

print(f"--- Αποτελέσματα για N = {N} ---")
print(f"Δειγματική Μέση Τιμή (Mean): {mean_est:.5f}")
print(f"Δειγματική Διασπορά (Variance): {var_est:.5f}")

--- Αποτελέσματα για N = 20000 ---
Δειγματική Μέση Τιμή (Mean): 0.04001
Δειγματική Διασπορά (Variance): 0.00000


1) Για Ν = 200 βρήκαμε μέση τιμή 0.0389
ενώ για Ν = 20000 βρήκαμε μέση τιμή 0.04.
2) Τα αποτελέσματα αυτά είναι ικανοποιητικά κοντά με την θεωρητική πιθανότητα μετάβασης μετά από 20 βήματα η οποία ήταν 0.04. Με αύξηση του N καταφέρνουμε καλύτερη προσέγγιση.
3) Η δειγματική διασπορά για Ν=200 τη βρήκαμε 0.00012 ενώ για Ν=20000 τη βρήκαμε 0.

H εκτίμηση δεν αλλάζει ουσιαστικά. Ο μέσος όρος για N = 20000 ξεκινώντας από την κατάσταση 3 βρέθηκε πρακτικά ίδιος με αυτόν ξεκινώντας από την κατάσταση 1. Μετά από 20 βήματα (που είναι θεωρητικά πολλά), το σύστημα έχει ξεχάσει από πού άρχισε ,λόγω της Μαρκοβιανής Ιδιότητας, και η πιθανότητα εξαρτάται μόνο από την αναλλοίωτη (στάσιμη) κατανομή, αφού είναι μη υποβιβάσιμη η αλυσίδα μας.


In [ ]:
!jupyter nbconvert --to html lab_2.ipynb

[NbConvertApp] Converting notebook lab_2.ipynb to html
[NbConvertApp] Writing 289639 bytes to lab_2.html
